### PR TITLE
LEG-142: Legislation document parser (EM / EN / IA / TN / CoP / OD)

### PR DESCRIPTION
**Pull request by QR-TNA**
Adds parsing of UK legislation associated documents (Explanatory Memoranda,
Explanatory Notes, Impact Assessments, Transposition Notes, Codes of Practice,
and Other Documents) from `.docx` to LegalDocML (AKN), with HTML rendering.
The bulk of the change is new code under `src/leg/` and fixtures under
`test/leg/`.

The branch is merged up to current `main` (v1.9.0). Leg parser output has been
QA'd separately via the leg fixture suite and a full EM/IA batch reparse, so
merging accepts that QA rather than relying on file-by-file review of the leg
tree.

## This PR does not change judgment or lawmaker output

Both are byte-identical to `main`:
- `git diff main -- test/judgments/` is empty
- `git diff main -- src/lawmaker/ test/lawmaker/` is empty

Every shared-code change is a bug fix or a no-op extension point that defaults
to main's behaviour and is overridden only by the legislation subclass.
Verified: judgment ApiTests and lawmaker tests pass with their fixtures
untouched from main.

---

## Reviewer guide: focus here (22 non-`src/leg/` files)

This PR is 573 files, but 546 are new leg code and fixtures under `src/leg/`
and `test/leg/`. Only these 22 files touch shared code or build config and
need review. Click through them as a checklist (the descriptive tables below
explain each):

Source and build (18):
[src/model2/Inline.cs](src/model2/Inline.cs) ·
[src/parsers/common/fields/Ref.cs](src/parsers/common/fields/Ref.cs) ·
[src/parsers/common/Inline2.cs](src/parsers/common/Inline2.cs) ·
[src/parsers/common/AlternateContent.cs](src/parsers/common/AlternateContent.cs) ·
[src/parsers/common/fields/Fields2.cs](src/parsers/common/fields/Fields2.cs) ·
[src/akn/Simplify.cs](src/akn/Simplify.cs) ·
[src/akn/Builder.cs](src/akn/Builder.cs) ·
[src/parsers/optimized/OptimzedParser.cs](src/parsers/optimized/OptimzedParser.cs) ·
[src/parsers/optimized/PreParser.cs](src/parsers/optimized/PreParser.cs) ·
[src/parsers/common/IDrawingResolver.cs](src/parsers/common/IDrawingResolver.cs) ·
[src/parsers/common/Inline.cs](src/parsers/common/Inline.cs) ·
[src/model2/Images.cs](src/model2/Images.cs) ·
[src/Program.cs](src/Program.cs) ·
[judgments.csproj](judgments.csproj) ·
[tna-judgments-parser.sln](tna-judgments-parser.sln) ·
[render-service/Program.cs](render-service/Program.cs) ·
[render-service/RenderService.csproj](render-service/RenderService.csproj) ·
[TRE/ASSOCIATED-DOCS-LAMBDA.md](TRE/ASSOCIATED-DOCS-LAMBDA.md)

Test (4):
[test/DocumentHelpers.cs](test/DocumentHelpers.cs) ·
[test/docx/TestClassifyHeading.cs](test/docx/TestClassifyHeading.cs) ·
[test/docx/TestStyleCascade.cs](test/docx/TestStyleCascade.cs) ·
[test/test.csproj](test/test.csproj)

### Genuine fixes 
| File | Change |
|---|---|
| `src/model2/Inline.cs` | Colour properties follow character-style-wins (CSS/Word specificity); other properties unchanged. Fixes black runs forced white inside coloured paragraph styles. 8 unit tests (`TestStyleCascade`). 0 judgment fixtures changed. |
| `src/parsers/common/fields/Ref.cs` | Null-safe bookmark parent, with a fall back to paragraph text instead of throwing on an un-numbered REF target. |
| `src/parsers/common/Inline2.cs` | Unknown inline elements degrade (recurse/innerText) instead of `throw`; OpenXML `rewritten://` hyperlink URIs are treated as un-emittable (keep text, drop the `a` wrapper), matching `fields/Hyperlink.cs`. |
| `src/parsers/common/AlternateContent.cs` | Handles `cx1` (Office 2016 Chart Extensions) via the fallback branch instead of throwing. |
| `src/parsers/common/fields/Fields2.cs` | FORMDROPDOWN field handling. Validated against a real IA document containing a dropdown when added; not captured as a committed test fixture. |

### Pure extension points (judgment and lawmaker byte-identical to main)
| File | Extension point (default) |
|---|---|
| `src/akn/Simplify.cs` | `ShouldStripStyleAttributes` (true = strip = main); `VisitText` made virtual |
| `src/akn/Builder.cs` | `DecorateBlockElement` (no-op); `ApplyTableCellStyleCleanup` (no-op); `AddHash` optional `prefix`/`localName` |
| `src/parsers/optimized/OptimzedParser.cs` | `ParseParagraphAndSubparagraphs` made virtual |
| `src/parsers/optimized/PreParser.cs` | `ReprocessFieldEndWithContent` (false = main TOC-end behaviour) |
| `src/parsers/common/IDrawingResolver.cs` (new) and `Inline.cs` | drawing-resolver hook; null for judgment falls back to `new WImageRef(...)`. Also removes a prior cross-namespace reference from shared `Inline.cs` to the leg `RenderSession` type. |
| `src/model2/Images.cs` | `WImageRef.Make()` additive factory (existing constructor untouched) |

### Leg infrastructure outside `src/leg/` (skim)
Leg-only by purpose; here because the build/CLI needs them.
| File | What |
|---|---|
| `src/Program.cs` | leg doc-type CLI dispatch plus `--output-html`, `--manifestation-name`, `--validate-akn` |
| `judgments.csproj` | leg embedded resources (subschemas, mapping CSV, XSL, CSS) |
| `tna-judgments-parser.sln`, `render-service/*` | leg chart-rendering microservice |
| `TRE/ASSOCIATED-DOCS-LAMBDA.md` | leg ops doc |

---

## Provenance: dual `ukm:Parser`

Leg AKNs emit two `ukm:Parser` entries (CLML schema change agreed with Jim:
repeating, optional `@Name`): `Name="core"` (shared `judgments.dll` version,
currently 1.9.0) and `Name="legislation"` (`src/leg/LegVersion.cs`). Bump
policy in `src/leg/VERSIONING.md`.

## Rendering (optional; the parser runs without it)

AKN is the primary output and is produced with no external tools: with no
renderer configured the parser uses a `NullRenderer` and parses normally.

On top of that, leg can render two things when the tools are present:
- **Chart / SmartArt images** embedded in the AKN, via LibreOffice (`soffice`).
  These are a real deliverable. They can be produced locally
  (`DocxToImageRenderer`) or, in the maintainer's setup, by a separate EC2
  LibreOffice instance reached over HTTP (`HttpRenderer` to `render-service/`).
- **HTML previews** of the AKN, via Saxon/Oxygen (`HtmlBuilder`). These are
  developer/test verification only, not a shipped artifact; the HTML snapshot
  tests `Skip` when Oxygen is absent.

Boundary notes for reviewers:
- The executable paths (`soffice`, Saxon/Java) come from deployment
  configuration (`SOFFICE_PATH`, `OXYGEN_HOME`), and all subprocesses use
  `UseShellExecute=false` with `ArgumentList` (no shell, no string-built
  commands). The Wiz OS-command-injection findings on these are false
  positives, cleared on the PR threads.
- `render-service/` is an optional, separately-deployable backend currently
  operated only by the maintainer. The parser does not require it; it is the
  scaled image-rendering path. Flagging the bus-factor of one for awareness.

## Testing

Judgment ApiTests, lawmaker, leg (EM/EN/IA/TN/CoP/OD plus HTML snapshots), and
helper unit tests pass. (Backlog E2E tests need LocalStack/S3 and a clean
working directory; unchanged from main.)

### CODE REVIEW

File: render-service/Program.cs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    private bool Probe() {
+        if (!File.Exists(config.SofficePath)) {
+            log.LogWarning("soffice not found at {Path}", config.SofficePath);
+            return false;
+        }
+        try {
+            var psi = new ProcessStartInfo {
+                FileName = config.SofficePath,
```
[the remark this answers is not in the document]
Comment (QR-TNA) [reply render-service/Program.cs:294]: #wiz_ignore False positive. Process runs with UseShellExecute=false and all args passed via ProcessStartInfo.ArgumentList (separate argv elements, never concatenated into a command line), so there's no shell to inject into. FileName is the soffice/java executable resolved from deployment config (SOFFICE_PATH / OXYGEN_HOME / LEG_SOFFICE_PATH), not user/request input; externally-influenced values (temp file paths, image-base) are passed as discrete args to a non-shell process. This is the secure pattern the guidance recommends.

File: render-service/Program.cs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                UseShellExecute = false,
+                CreateNoWindow = true,
+            };
+            psi.ArgumentList.Add("--version");
+            using var p = Process.Start(psi);
```
[the remark this answers is not in the document]
Comment (QR-TNA) [reply render-service/Program.cs:301]: #wiz_ignore False positive. Process runs with UseShellExecute=false and all args passed via ProcessStartInfo.ArgumentList (separate argv elements, never concatenated into a command line), so there's no shell to inject into. FileName is the soffice/java executable resolved from deployment config (SOFFICE_PATH / OXYGEN_HOME / LEG_SOFFICE_PATH), not user/request input; externally-influenced values (temp file paths, image-base) are passed as discrete args to a non-shell process. This is the secure pattern the guidance recommends.

File: src/leg/HtmlBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        string cssPath = Path.Combine(tempDir, "associated-docs.css");
+        string htmlPath = Path.Combine(tempDir, "out.html");
+
+        try {
+            akn.Save(aknPath);
+            ExtractEmbeddedResource("leg.akn2html.xsl", xslPath);
+            ExtractEmbeddedResource("leg.associated-docs.css", cssPath);
+
+            var psi = new ProcessStartInfo {
+                FileName = javaExe,
```
[the remark this answers is not in the document]
Comment (QR-TNA) [reply src/leg/HtmlBuilder.cs:55]: #wiz_ignore False positive. Process runs with UseShellExecute=false and all args passed via ProcessStartInfo.ArgumentList (separate argv elements, never concatenated into a command line), so there's no shell to inject into. FileName is the soffice/java executable resolved from deployment config (SOFFICE_PATH / OXYGEN_HOME / LEG_SOFFICE_PATH), not user/request input; externally-influenced values (temp file paths, image-base) are passed as discrete args to a non-shell process. This is the secure pattern the guidance recommends.

File: src/leg/HtmlBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            };
+            psi.ArgumentList.Add("-cp");
+            psi.ArgumentList.Add(libGlob);
+            psi.ArgumentList.Add(SaxonMainClass);
+            psi.ArgumentList.Add($"-xsl:{xslPath}");
+            psi.ArgumentList.Add($"-s:{aknPath}");
+            psi.ArgumentList.Add($"-o:{htmlPath}");
+            psi.ArgumentList.Add($"image-base={imageBase}");
+
+            using var process = Process.Start(psi);
```
[the remark this answers is not in the document]
Comment (QR-TNA) [reply src/leg/HtmlBuilder.cs:69]: #wiz_ignore False positive. Process runs with UseShellExecute=false and all args passed via ProcessStartInfo.ArgumentList (separate argv elements, never concatenated into a command line), so there's no shell to inject into. FileName is the soffice/java executable resolved from deployment config (SOFFICE_PATH / OXYGEN_HOME / LEG_SOFFICE_PATH), not user/request input; externally-influenced values (temp file paths, image-base) are passed as discrete args to a non-shell process. This is the secure pattern the guidance recommends.

File: src/leg/common/rendering/DocxToImageRenderer.cs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    }
+
+    private bool ConvertToHtml(string inputDocx, string outDir, CancellationToken ct) {
+        // Per-invocation user profile so concurrent soffice processes don't collide on the default profile.
+        string profileDir = Path.Combine(outDir, "profile");
+        Directory.CreateDirectory(profileDir);
+        string profileUri = new Uri(profileDir).AbsoluteUri;
+
+        var psi = new ProcessStartInfo {
+            FileName = sofficePath,
```
[the remark this answers is not in the document]
Comment (QR-TNA) [reply src/leg/common/rendering/DocxToImageRenderer.cs:84]: #wiz_ignore False positive. Process runs with UseShellExecute=false and all args passed via ProcessStartInfo.ArgumentList (separate argv elements, never concatenated into a command line), so there's no shell to inject into. FileName is the soffice/java executable resolved from deployment config (SOFFICE_PATH / OXYGEN_HOME / LEG_SOFFICE_PATH), not user/request input; externally-influenced values (temp file paths, image-base) are passed as discrete args to a non-shell process. This is the secure pattern the guidance recommends.

File: src/leg/common/rendering/DocxToImageRenderer.cs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        psi.ArgumentList.Add("--headless");
+        psi.ArgumentList.Add("--convert-to");
+        psi.ArgumentList.Add("html");
+        psi.ArgumentList.Add("--outdir");
+        psi.ArgumentList.Add(outDir);
+        psi.ArgumentList.Add(inputDocx);
+
+        subprocessGate.Wait(ct);
+        try {
+            using var proc = Process.Start(psi);
```
[the remark this answers is not in the document]
Comment (QR-TNA) [reply src/leg/common/rendering/DocxToImageRenderer.cs:100]: #wiz_ignore False positive. Process runs with UseShellExecute=false and all args passed via ProcessStartInfo.ArgumentList (separate argv elements, never concatenated into a command line), so there's no shell to inject into. FileName is the soffice/java executable resolved from deployment config (SOFFICE_PATH / OXYGEN_HOME / LEG_SOFFICE_PATH), not user/request input; externally-influenced values (temp file paths, image-base) are passed as discrete args to a non-shell process. This is the secure pattern the guidance recommends.

**Comment by jurisdatum** [comment]
Thanks — the reviewer guide makes a PR this size approachable, and pointing straight at the shared-code surface is exactly what I needed. Started digging in; I'll follow up properly before long.

File: src/Program.cs [hunk 7]
```diff
@@ -95,6 +135,8 @@ static int Transform(ParseResult parseResult)
             return Success;
         } else {
             logger?.LogCritical("unrecognized document type: {}", hint);
+            Console.Error.WriteLine($"Error: Invalid hint '{hint}'. Supported values: 'em', 'en', 'ia', 'tn', 'cop', 'od', or a Lawmaker type such as 'nipubb', 'uksi', or 'ukprib'.");
+            Environment.Exit(1);
```
Comment (jurisdatum) [inline src/Program.cs:148]: I think this `else` accidentally removes the old no-hint judgment CLI path. On `main`, running with no `--hint` fell through to the judgment `Api.Parser` block just below; here `DocNames.GetDocName(hint)` returns null for an empty hint, so `Environment.Exit(1)` makes lines 141–162 unreachable. A plain `--input foo.docx` now exits 1 instead of parsing as a judgment.

The fixture/API tests wouldn't catch this since they don't go through CLI dispatch. Did you mean to drop the no-hint judgment path, or should an empty hint still fall through like before? Unknown hints becoming invalid may be fine; the bit that reads as a pre-merge regression is the empty/no-hint caselaw path.

File: src/leg/common/BaseMetadata.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    public Tuple<string, int> AltNum { get; init; }
+
+    internal static BaseMetadata Make(List<IBlock> header, WordprocessingDocument doc, LegislativeDocumentConfig config) {
+        string name = BaseHeaderSplitter.GetDocumentType(header, config);
+        // If header parsing fails to identify document type, use the default
+        if (string.IsNullOrEmpty(name)) {
+            name = config.DefaultDocumentType;
+        }
+        string number = BaseHeaderSplitter.GetDocumentNumber(header);
+        string uri = number is null ? null : RegulationNumber.MakeURI(number) + config.UriSuffix;
```
Comment (jurisdatum) [inline src/leg/common/BaseMetadata.cs]: Small base-class edge case: this guards `number is null`, but `RegulationNumber.MakeURI(number)` can also return `null` for draft/unassigned numbers that `RegulationNumber.Is` accepts. If this fallback metadata path sees one of those values, `MakeURI(number) + config.UriSuffix` will produce a bare suffix such as `"/other-document"` because string concatenation treats the null as empty.

The concrete associated-doc parsers appear to use their document-specific metadata classes today, so this may be defensive rather than fixture-affecting. But if `BaseMetadata` remains as the fallback, it probably wants to compute the base URI first and append the suffix only when that is non-null.

File: src/leg/common/BaseHelper.cs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        Config = config;
+    }
+
+    public IXmlDocument Parse(
+        Stream docx, bool simplify = true, string filename = null,
+        string manifestationName = Builder.DefaultManifestationName,
+        bool allowUnrenderedCharts = true,
+        IDrawingRenderer renderer = null) {
+        WordprocessingDocument word = UK.Gov.Legislation.Judgments.AkomaNtoso.Parser.Read(docx);
+        return Parse(word, simplify, filename, manifestationName, allowUnrenderedCharts, renderer, docxBytes: null);
```
Comment (jurisdatum) [inline src/leg/common/BaseHelper.cs]: Small API footgun, unless I’m missing something: this overload accepts a `renderer`, but then calls the shared parse path with `docxBytes: null`. Since `RenderSession` needs the original `.docx` bytes to render charts/SmartArt, a caller using `Parse(Stream, ..., renderer)` would not actually get rendered drawings from that renderer. The `byte[]` overload does work.

Could the stream overload buffer to bytes and delegate to the byte[] overload, or otherwise avoid exposing `renderer` here?

**Review by jurisdatum — COMMENTED** [review]
Thanks for walking me through this on the call. The boundary mostly checks out: judgment and lawmaker fixture outputs come through byte-identical to `main`, and the shared-code hooks I looked at default to the existing behaviour.

Following the reviewer guide, I haven't tried to review the whole leg fixture corpus. I focused on the shared core and the system boundaries. I've left three inline comments:

- **One I'd treat as pre-merge:** the no-hint judgment CLI path in `Program.cs` looks like it regressed. A plain caselaw CLI parse with no `--hint` now exits before reaching the judgment parser. The fixture/API tests don't catch it because they don't go through CLI hint dispatch.
- **Two smaller, code-anchored things:** a draft-number URI edge case in `BaseMetadata`, and a latent API footgun where the `Stream` parse overloads accept a `renderer` they can't actually use. Both are non-blocking from my point of view — fix or wave off as you see fit.

The larger questions we talked about — whether leg should compile into the shared `judgments.dll`, the eventual common/judgments/lawmaker/legislation split, and where the LibreOffice/Oxygen tooling should live — I don't think belong in this PR thread. They're really a cross-team conversation, and I'll follow up on that separately rather than gate the PR on them.

Great work. Hope to see associated documents in HTML on legislation.gov.uk very soon!

File: src/Program.cs (hunk 7 above)
Comment (QR-TNA) [reply src/Program.cs:148]: Thanks for spotting this. Fixed now. The no-hint (and empty-hint) case now falls through to the judgment Api.Parser path as it did on main. 

One extra thing I found while fixing it: DocNames.GetDocName doesn't return null for a non-empty unrecognised hint, it throws, so the friendly "Invalid hint" message was unreachable and a bad hint produced a raw stack trace. The dispatch now catches that, so an unknown hint gives the intended error and exits 1, while no hint parses as a judgment. Verified both paths end to end via the CLI.

File: src/leg/common/BaseMetadata.cs (hunk 8 above)
Comment (QR-TNA) [reply src/leg/common/BaseMetadata.cs]: Agreed. It now computes the base URI first and only appends UriSuffix when that is non-null, so a draft/unassigned number no longer yields a bare suffix like /other-document.

File: src/leg/common/BaseHelper.cs (hunk 9 above)
Comment (QR-TNA) [reply src/leg/common/BaseHelper.cs]: Fixed: the Stream overload now buffers to bytes and delegates to the byte[] overload, so a caller passing a renderer actually gets rendered drawings.  Confirmed against RenderSession.GetRenderedDrawing, which no-ops when DocxBytes is null regardless of the renderer.

**Review by jurisdatum — APPROVED** [review]
Looks good
